sign/bls/bls.go

```diff
@@ -0,0 +1,414 @@
+		kk := (interface{})(k.key).(G2)
+		return !kk.g.IsIdentity() && kk.g.IsOnG2()
+	default:
+		panic(ErrInvalid)
+	}
+}
+
+func (k *PublicKey[K]) Equal(x crypto.PublicKey) bool {
+	xx, ok := x.(*PublicKey[K])
+	if !ok {
+		return false
+	}
+
+	switch (interface{})(k).(type) {
+	case *PublicKey[G1]:
+		xxx := (interface{})(xx.key).(G1)
+		kk := (interface{})(k.key).(G1)
+		return kk.g.IsEqual(&xxx.g)
+	case *PublicKey[G2]:
+		xxx := (interface{})(xx.key).(G2)
+		kk := (interface{})(k.key).(G2)
+		return kk.g.IsEqual(&xxx.g)
+	default:
+		panic(ErrInvalid)
+	}
+}
+
+// MarshalBinary returns a slice with the compressed
+// representation of the underlying element in G1 or G2.
+func (k *PublicKey[K]) MarshalBinary() ([]byte, error) {
+	switch (interface{})(k).(type) {
+	case *PublicKey[G1]:
+		kk := (interface{})(k.key).(G1)
+		return kk.g.BytesCompressed(), nil
+	case *PublicKey[G2]:
+		kk := (interface{})(k.key).(G2)
+		return kk.g.BytesCompressed(), nil
+	default:
+		panic(ErrInvalid)
+	}
+}
+
+func (k *PublicKey[K]) UnmarshalBinary(data []byte) error {
+	switch (interface{})(k).(type) {
+	case *PublicKey[G1]:
+		kk := (interface{})(&k.key).(*G1)
+		return kk.setBytes(data)
+	case *PublicKey[G2]:
+		kk := (interface{})(&k.key).(*G2)
+		return kk.setBytes(data)
+	default:
+		panic(ErrInvalid)
+	}
+}
+
+// KeyGen derives a private key for the specified group (G1 or G2).
+// The length of ikm material should be at least 32 bytes length.
+// The salt value should be either empty or a uniformly random
+// bytes whose length equals the output length of SHA-256.
+func KeyGen[K KeyGroup](ikm, salt, keyInfo []byte) (*PrivateKey[K], error) {
+	// Implements recommended method at:
+	// https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-05#name-keygen
+	if len(ikm) < 32 {
+		return nil, ErrShortIKM
+	}
+
+	ikmZero := make([]byte, len(ikm)+1)
+	keyInfoTwo := make([]byte, len(keyInfo)+2)
+	copy(ikmZero, ikm)
+	copy(keyInfoTwo, keyInfo)
+	const L = uint16(48)
+	binary.BigEndian.PutUint16(keyInfoTwo[len(keyInfo):], L)
+	OKM := make([]byte, L)
+
+	var ss GG.Scalar
+	for tries := 8; tries > 0; tries-- {
+		rd := hkdf.New(sha256.New, ikmZero, salt, keyInfoTwo)
+		n, err := io.ReadFull(rd, OKM)
+		if n != len(OKM) || err != nil {
+			return nil, err
+		}
+
+		ss.SetBytes(OKM)
+
+		if ss.IsZero() == 1 {
+			digest := sha256.Sum256(salt)
+			salt = digest[:]
+		} else {
+			return &PrivateKey[K]{key: ss, pub: nil}, nil
+		}
+	}
+
+	return nil, ErrKeyGen
+}
+
+// Sign computes a signature of a message using a key (defined in
+// G1 or G1).
+func Sign[K KeyGroup](k *PrivateKey[K], msg []byte) Signature {
+	if !k.Validate() {
+		panic(ErrInvalidKey)
+	}
+
+	switch (interface{})(k).(type) {
+	case *PrivateKey[G1]:
+		var Q GG.G2
+		Q.Hash(msg, []byte(dstG2))
+		Q.ScalarMult(&k.key, &Q)
+		return Q.BytesCompressed()
+	case *PrivateKey[G2]:
+		var Q GG.G1
+		Q.Hash(msg, []byte(dstG1))
+		Q.ScalarMult(&k.key, &Q)
+		return Q.BytesCompressed()
```

Comment:
The compression matches the spec also in the corner cases?

Reply:
all the points have fixed-size length. or do you mean something different?

Reply:
Case of point at infinity.

Reply:
Tests for serialization here depend only on the unique ciphersuite: BLS12-381 curve
The encoding format is tested in the ecc/bls12381 package, see: https://github.com/cloudflare/circl/blob/main/ecc/bls12381/encoding_test.go#L73